extensions/narayana-jta/deployment/src/main/java/io/quarkus/narayana/jta/deployment/NarayanaJtaProcessor.java

```diff
@@ -135,6 +147,16 @@ public void build(NarayanaJtaRecorder recorder,
         recorder.setConfig(transactions);
     }
 
+    @BuildStep
+    @Record(RUNTIME_INIT)
+    @Consume(NarayanaInitBuildItem.class)
+    @Consume(SyntheticBeansRuntimeInitBuildItem.class)
+    public void startRecoveryService(NarayanaJtaRecorder recorder,
+            List<JdbcDataSourceBuildItem> jdbcDataSourceBuildItems, TransactionManagerConfiguration transactions) {
+        jdbcDataSourceBuildItems.forEach(i -> recorder.addDataSource(i.isDefault(), i.getName()));
```

Comment:
I don't like this pattern, you are keeping datasources in non-transient recorder member field and query them later, why don't you pass it in when you call `startRecoveryService`? you are doing both at runtime anyway

Reply:
Good question! I recall that `JdbcDataSourceBuildItem` can only be availiable at runtime. If we just call something like `recorder.startRecoveryService(transactions, jdbcDataSources)`, it seems not working. Anyway, I will check it again.

Reply:
I don't doubt it as it comes from deployment dep, but you can pass in Map<Boolean, String> isDefaultToDsName (choose better name)

Reply:
Thanks for the hint!

Reply:
Hi @michalvavrik,

I tried the following codes
```
List<Pair<Boolean, String>> dataSourcesNames = new ArrayList<>();
jdbcDataSourceBuildItems.forEach(i -> dataSourcesNames.add(Pair.of(i.isDefault(), i.getName())));

recorder.startRecoveryService(transactions, dataSourcesNames);
```

But it does not work since the application bootstrap fails with
```
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: java.lang.NullPointerException: Cannot invoke "io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.doLoad(io.quarkus.deployment.recording.BytecodeRecorderImpl$MethodContext, io.quarkus.gizmo.MethodCreator, io.quarkus.gizmo.ResultHandle)" because "parameter" is null
```

It seems that we can not pass a new created value at build time. Or we have to get a RuntimeValue from the recorder, I recall that it could be the reason I used `recorder.addDataSource()` before.

Reply:
sure @zhfeng , that is why I suggested `Map<Boolean, String>` and not `List<Pair<Boolean, String>>`, the reference guide https://quarkus.io/guides/writing-extensions#bytecode-recording you can see that Map, primitive types and string are supported, but `org.apache.commons.lang3.tuple.Pair` would need to be recorded at runtime.

So if you insist on `List<Pair<Boolean, String>>`, you can achieve that easily by adding to recorder

```
public Pair<Boolean, String> toPair(String dsName, boolean isDefault) {
  return Pair.of(isDefault, dsName);
}
```
and in processor doing

```
jdbcDataSourceBuildItems.forEach(i -> dataSourcesNames.add(recorder.toPair(i.isDefault(), i.getName())));
```

I don't want to push you too much is this is not really issue, but in general I think it is better not to hold these values on recorder for whole app running when you only need to pass them to recorded method. Choice is yours :-)

Reply:
Oh, interesting!. It refresh my knowledge about bytecode recording. Thanks!

Reply:
And yeah, it could be better to try to pass with `Map<Boolean, String>` and remove dependency on `org.apache.commons:commons-lang3`.

Reply:
> And yeah, it could be better to try to pass with `Map<Boolean, String>` and remove dependency on `org.apache.commons:commons-lang3`.

that's what I thought but didn't want to badger too much!

Reply:
I see :)